### ISSUE TITLE
Full guide on how to deploy Mailu on k8s - sponsorship

### ISSUE THREAD

**Issue by arturhefczyc** [Mailu/helm-charts#118]
Hi,

On a surface using helm charts allows for a simple deployment but there are many pitfalls such as shared volume, ports forwarding, certificates. And details are scarces and difficult to get it all together.

So, it would be very useful for people to have a complete guide on how to set everything up to have a working Mailu deployment.

I am offering a sponsorship for preparing such a guide or even 2 guides. One for managed k8s like DigitalOcean and one for home lab network running on RPis or other small servers with k3s.

Such a guide should have example config for shared volumes, ports forwarding config on LB and firewall settings, certificates auto renewal setup, sealed secrets, backups, restore. I mean a complete working system.

I posted a bounty for the issue: https://www.bountysource.com/issues/103130194-full-guide-on-how-to-deploy-mailu-on-k8s-sponsorship

So, here are my requirements for such a guide:

This should be “HOWTO” kind of guide to show how to get fully functional system up and running.

1. Mailu running on managed k8s system, preferably Oracle Cloud
2. Mailu running on cluster of 3 nodes
3. WIth automated backups and instructions on how to recover from backup (maybe velero?)
4. With automatically renewed certificates
5. Using sealed secrets
6. Setup controlled by fluxcd
7. Guide should also cover LB configuration for receiving mails and connecting via IMAP
8. Sending emails from own server is always a problem so the guide may also include settings on how to use external SMTP service. However this is already covered in Mailu docs anyway, so this part can be very brief like an example config file and how to apply these custom configs into the k8s system.

Any more suggestions are welcome.

**Comment by arturhefczyc** [comment]
Since I posted a bounty I have also added more details about the guide to the main issue description.

**Comment by micw** [comment]
Hello,
thank you very much in investing in Mailu's documentation. I appreciate this very much!

After reading your bounty, I think you should be more clear in your expectations (e.g. should it one or two guides. which of the extras [backup, sealed secrets, fluxcd] are mantadory, ...).

I personally see it more important to have clear documentation, in which exact environments mailu can run (and in which it cannot) than having the additional stuff like sealed secrets and a certain CD.

Kind regards,
Michael.

**Comment by arturhefczyc** [comment]
Thank you for your comments and suggestion. 

I also like to have a good quality documentation. However, Mailu is very complex system in itself and hosting email server is a complex subject too, plus k8s is not a simple thing either.

Even if you have a very good documentation for these 3 elements, it might be hard to put all these together.
Therefore I find a HOWTO kind of guide invaluable. At least for me, this is how I learn. I first like to have a working setup and then dig into documentation to customize it. Hence my bounty to have the guide to put things together.

**Comment by zestysoft** [comment]
I've thought about providing howto's like this, however what always stops me is one of the problems of using kubernetes -- it's made up of a lot of parts and there are many different implementations of those.  Take these as an example:

- What version of kubernetes should this be written for? -- lots of breaking changes depending on which version is running, some versions don't work with some of the implementations (for instance Kasten doesn't work with 1.22.x yet).
- kubernetes setup: kubeadm vs minikube vs kops vs microk8s vs ...
- networking: calico vs flannel vs weave vs ...
- load balancers: metallb vs purelb vs openelb vs ...
- storage: ceph vs gluster vs samba vs nfs vs ...
- backup:  velero vs kasten vs cohesity vs portwrox vs ...
- hardware of nodes:  amd64 vs arm64
- OS running on nodes: ubuntu vs redhat vs ...
- gateway device for port forwarding: pfsense, opensense, untable, openwrt, ubiquiti ...
- HA solutions: BGP vs OSPF
- Ingress ...
- TLS certificate generation ...
- Secret management (not the default base64)

To produce this howto, you'd also need to produce howto's for all of the parts, and possible all of the different implementations of those parts.  That or assume the reader is willing to drop their implementation choices and start over with what I happen to have chosen for myself.

Maybe scoping it down to a single managed solution (DigitalOcean?) might help, assuming all of these implementation choices are picked for you by it.

**Comment by lazaridiscom** [comment]
> scope of this guide: for medium level technical people who know their way around
> managed, 3-node k8s cluster.
> only free and open source tools should be used.
> providers would be in the following order:
> 
>     1. Oracle managed k8s
>     2. DigitalOcean or Linode managed k8s
>     3. OvhCloud

@arturhefczyc, I'd be interested to provide this, though not via bountysource (which has a highly amateurish manual payout process). If you feel that a high-experienced hw/sw-designer (who wants to come in contact with kybernetes) can execute this task within 5 focused full-time work-days, please contact me via info @ my-username.

I would simply do the setup as a newcomer, and document compactly the overall process, trying to produce as many "executable" installation steps (e.g. cli instructions).

**Comment by arturhefczyc** [comment]
@lazaridiscom I have already paid to bountysource for the sponsorhip of this issue. If I hire you to do this as a contract work, it would be double payment for me.
Let's stick to the bountysource. I have used it in the past and my experience with it is positive.

**Comment by lazaridiscom** [comment]
Bountysource is ok, but when it comes to cash-out, it's an amateurish disaster, which makes getting closure and go on with the next task quite difficult.

My personal record is 5 (I had to send 5 emails within 2 weeks in order to finally get the amount to my paypal). 

I mean, what kind of a circus-show is this:

```
I submitted a cash out request. When will I be paid?

Cash out requests are processed manually by a volunteer-run team. Generally speaking, cash outs are run on Friday or over the weekend. We aim to process all requests submitted before the start of Friday (Pacific Time).

Note: The timing here can vary substantially based on volunteer availability. If your request is still pending after a week or two please send us a support request and poke it occasionally until somebody can get to it
```

https://github.com/bountysource/core/wiki/Frequently-Asked-Questions#i-submitted-a-cash-out-request-when-will-i-be-paid

-

So: no way i create again an account with those amateurs.

Note that this is not a  "trivial documentation-task", is a "simplification/integration/documentation task". 

If you cant find anyone else, note that you can cash-out any paid bounties (loosing the 10%).

**Comment by arturhefczyc** [comment]
As I was always on the other end of the Bountysource I wasn't aware of the payment issues. I used it a few times and never seen any complains. 
Assumed that they get cut from all transactions plus interest from money on their account and this is enough for them to have staff running the service. Looks like I assumed wrong.

Do you know of any alternative you can recommend in place of Bountysource?

Anyway, money is now on Bountysource so for this task I do not want to change it. I could possibly offer more if that would be something to convince someone.
On the other hand, it looks like I am getting close myself to get everything working they way I need. Once I have the setup up and running with all the issues resolved I am going to write the guide myself. Then, the bounty would be reassigned for something else.

**Comment by lazaridiscom** [comment]
They get 10%.

Our workaround here would be that:

- I execute the task
- you claim & payout the bounty to yourself
- send me the payed-out amount via usual paypal.

as for
> On the other hand, it looks like I am getting close myself to get everything working they way I need. Once I have the setup up and running with all the issues resolved I am going to write the guide myself. 

Don't underestimate the effort to write a guide after(!) doing the work. Remember to take many notes, ideally cli-commands. Or to write the guide during your work.

In any way, I'm available to execute this over the holidays, with the usual guideline "Best Documentation is none at all" (e.g.: try to eliminate all manual steps, thus the documentation shrinks down significantly)

> Do you know of any alternative you can recommend in place of Bountysource?

I noticed a few, but never tried one of those.

**Comment by DAlperin** [comment]
Hi @arturhefczyc! Can I take a crack at this?

**Comment by DAlperin** [comment]
Here are my thoughts:

> Mailu running on managed k8s system, preferably DigitalOcean

Does Linode work? Understand if not but I am already in the process of doing a bunch with Linodes managed kubernetes.

> WIth automated backups and instructions on how to recover from backup (maybe velero?)

Velero works. I will warn you that it does add a not insignificant amount of complexity to the system since it requires an object storage system (s3, azure storage, minio, etc)

> With automatically renewed certificates

The chart supports this, I can explain how to configure it

> Using sealed secrets
> Setup controlled by fluxcd

I can use fluxcds integration with sealed-secrets

My idea for the deliverable of the tutorial would be a fluxcd repository (I will release a working sample one).
The repo would contain a configuration file for the mailu chart, fluxcd chart, and sealed secrets chart.

Let me know if this works for you, if so I can have it done by early next week.

**Comment by arturhefczyc** [comment]
Hi,

Thank you for your interest. Yes, please go ahead.
1. Linode works just as well as any other provider
2. It does not have to be velero but regular, automated backups is a must have. I guess you could use Linode's object storage if necessary. It does not matter. This is supposed just to be a TODO how to get it working. It might be a good idea to run Mailu on one provider (Linode) and keep backups on another provider (AWS). But we do not need this kind of complication for this guide.
3. fluxcd repository as a tutorial would work perfectly well but it needs to have a good README as a documentation.

Please go ahead and let me know if you have any questions or if you need anything.

**Comment by DAlperin** [comment]
@arturhefczyc Almost done with my tutorial, had a quick question. As I write up all the steps I can either give the 'idiomatic' fluxcd way, running `flux create` commands that abstract away whats really happening, or I can explain each manifest. I am kinda partial to showing and explaining each manifest since I feel like it gives people more insight into how it works, but I don't really mind either way.

**Comment by DAlperin** [comment]
(some things like creating fluxcd source manifests I'd keep as the commands since they generate boilerplate not intended to be modified or read)

**Comment by arturhefczyc** [comment]
There is no need to explain each manifest unless there is really something special or interesting to point out. Also, if there are some possible configuration settings required in certain cases, it would be good to mention them.
But if there are any modifications, customizations or custom values then it definitely should be explained.

In general, anything non-standard, deviations from defaults should be explained.

**Comment by DAlperin** [comment]
Hey @arturhefczyc. Just wanted to give you a quick update on where things stand. 

Part of the reason running mailu (and all mailservers) on kubernetes is so hard is because mailu does not configure postfix and dovecot to accept the proxy protocol (this allows the mail servers to know what the actual client ip is, un-obscured by the ingress). Unfortunately that is a requirement in this case since no managed k8s service I have ever seen (of the major 10 or so) allows you to assign a static ip to a deployment, only and ingress (for people running their own cluster this is not an issue since they directly control, at least in theory, their nodes and their network).

Long story short I have ended up forking mailu and the helm charts with proxy protocol support that I will release along with the tutorial for people to use until I can get stuff upstreamed (I can't find the issue # but I think this will close a bug in https://github.com/mailu/mailu requesting proxy-protocol support)

I am a little worried this tutorial is going to be a bit of a monster due to the inherent difficulty with both mail servers and kubernetes. Comes with the territory I guess.

**Comment by arturhefczyc** [comment]
The complexity of the subject is/was one of the main reasons I offered bounty for this.
On the other hand, I might underestimated the effort required for the work, therefore, I can offer an extra bonus after completion of the task.

**Comment by OnekO** [comment]
The guide you are looking for is quite extensive, as there are a few items that would require a full guide for themselves (velero,flux,LB on diferents providers, secrets management...). If I could help with my experience deploying Mailu in my baremetal k8s with helm, I'll do for free, just spend the money on mailu team. But for sure it'll be shorter than the required one on bountysource.

**Comment by arturhefczyc** [comment]
> The guide you are looking for is quite extensive, as there are a few items that would require a full guide for themselves (velero,flux,LB on diferents providers, secrets management...). If I could help with my experience deploying Mailu in my baremetal k8s with helm, I'll do for free, just spend the money on mailu team. But for sure it'll be shorter than the required one on bountysource.

I am not looking for a generic guide for all possible platforms and environments. That would be too much for sure. I am looking for an example HOWTO on 1 specific k8s managed platform with some specific conditions.
There are already guides covering secrets management, velero, flux. So I would just skip over this, maybe pointing to online documentation. Just provide an example, working manifest files.

In other words, I am looking for a guide: "How to get fully working Mailu on k8s on a cloud using a particular provider.".

Setting up Mailu on home k8s is doable but not really reliable approach, since home network is not reliable enough. This is why I am looking for a guide to set it up in the cloud.

**Comment by abebeos** [comment]
Re Bountysource, please be aware of this:

https://github.com/abebeos/main/issues/2

**Comment by abebeos** [comment]
@arturhefczyc See update on bountysource, I can start work on this.

**Comment by arturhefczyc** [comment]
Great, I increased bounty to requested.

**Comment by abebeos** [comment]
All good then, starting today.

**Comment by arturhefczyc** [comment]
Let me know if you need anything. Perhaps I could give you access to our k8s cluster on Oracle cloud?

**Comment by abebeos** [comment]
I'll let you know, for now I'll stay "off-cloud". 

Is `oracle cloud` your no. 1 choice?

**Comment by arturhefczyc** [comment]
Yes, oracle cloud is my 1. choice after trying several other options.

**Comment by abebeos** [comment]
Ok, will go then with Oracle, and have DigitalOcean as a fallback.

**Comment by arturhefczyc** [comment]
Ok, just as you know. I have a really bad experience with DigitalOcean.

Had a paid account with them. Used it to learn and experiment with k8s and planned to use it as a k8s cluster to host all my stuff. Nothing fancy or weird. Hosting email, some static websites, and other normal services.
Everything worked great, so upgraded to higher level cluster. Then, some day, I got a message from them that my account has been locked for breaking their rules.
In the message they said they cannot disclose what rules I broke and what exactly wrong I did. I appealed their decision only to receive the same response. They won't tell what I did wrong and what rules I broke and the account remains locked.
On top of this, I still had cluster running with them which they would not stop, despite locking my account. I could not stop these services either because I had no access to account. So they were happily charging me for all the stuff running on my account. Mind, I have just upgraded k8s cluster to higher level.
The only good thing, I did not have any production systems running there yet. I just finished my tests and was getting ready to deploy real stuff.

So, this is just a word of caution before you consider DO for any real stuff.

**Comment by abebeos** [comment]
> In the message they said they cannot disclose what rules I broke and what exactly wrong I did. I appealed their decision only to receive the same response. They won't tell what I did wrong and what rules I broke and the account remains locked.

Oh my, this is actually enough reason to not use DO in any case.

Unbelievable, IT-landscape gets worse and worse.

**Comment by abebeos** [comment]
@arturhefczyc , please see your invite here on github.

**Comment by abebeos** [comment]
@vibhuti019 , i'm currently active on this task/bounty, and will publish most possible within the next days the initial documentation.

**Comment by abebeos** [comment]
For anyone interested to 'replay' the installation (or to just help out with some domain-knowledge), 'beta-testing' (kind of) will start tomorrow. Focus is Oracle Cloud k8s ("Quick Create" cluster with all defaults, accesses via public node). Local env is ubuntu 20.04 (native or virtual), but any other bash-capable os should do it).

**Comment by abebeos** [comment]
> For anyone interested to 'replay' the installation

Please see: https://github.com/abebeos/mailu-k8s

**Comment by abebeos** [comment]
@vibhuti019 , @all, my goal here was to achieve a professional-grade setup on a managed k8s. Sadly, I burned myself out, it looks like the problem with the load-balancer is not solvable without code-level work (which would be a bounty in the $10k region, so, beyond the scope of this integration-configuration- documentation bounty)

So, the bounty is "open" again (technically it was all the time, just making it clear that it's ok if you want to start work on this).

I may continue at some point to document the (theoretically) simpler local setup.

**Comment by DAlperin** [comment]
@abebeos this is where I got stuck when I tried this a little bit ago. Maybe at some point I'll clean up some patches I started on nginx to maybe make this a bit easier for whoever comes next.

**Comment by vibhuti019** [comment]
@abebeos I am starting to work on this, can you please brief how you tried to takle the issue and the problem occurred, it would help me and even everyone else working.

**Comment by abebeos** [comment]
@vibhuti019 , Just read through this issue, you'll find your way. Please note that this comes down to code-level work, and is most possibly not solvable without the support of the involved teams. 

@DAlperin , as I assess it, this looks like work to be done by the core-team-members. But you can try to publish your stuff of course.

**Comment by abebeos** [comment]
> Long story short I have ended up forking mailu and the helm charts with proxy protocol support that I will release along with the tutorial for people to use until I can get stuff upstreamed (I can't find the issue # but I think this will close a bug in https://github.com/mailu/mailu requesting proxy-protocol support)

@DAlperin , could you provide a link to those repositories?

**Comment by DAlperin** [comment]
@abebeos, I never made them public. I can try to clean them up a little and release them later this week. As you have discovered, it unfortunately does not solve the fundamental networking issues that we run up against with managed k8s. Like I said, I did play around with patching nginx to make it behave better with k8s and mail servers but I ultimately could not justify the time it would take to finish, nor the time it would take to get everything upstreamed. (I am not also convinced it would have been a complete solve anyway)

I'm sorry I couldn't be of more help. But like I said I'll do my best to clean up some of my forks into releasable states... If anyone who is taking a crack at this wants to chat specifically/ask questions about the networking limitations specifically shoot me an email at dov(at)dov(dot)dev and I'll see if I can dig up any of my notes to help.

I applaud whoever is the one to figure this out eventually. It has claimed an absurd number of engineer-hours thus far. We might need to start keeping track of our own [`total_hours_wasted_here`](https://stackoverflow.com/a/482129) counter ;)

**Comment by abebeos** [comment]
@DAlperin, could you assess if this here https://github.com/Mailu/Mailu/pull/2300 would solve the problem, or would there be much more to it?

> nor the time it would take to get everything upstreamed.

Sadly, this is sometimes the highest effort of all.

> It has claimed an absurd number of engineer-hours thus far.

Well, on my side the "trial&error"/"plumbing" hours are at an absurd level, too, and much higher than the engineering hours.

Though possibly this is solvable easier if we team-up.

@arturhefczyc maybe you want to apply a separate bounty to https://github.com/Mailu/helm-charts/issues/172 to solve this specific "managed k8s problem"? work on https://github.com/abebeos/mailu-k8s has isolated/documented this problem, but fixing it feels quite beyond the scope. @DAlperin and some others have done some work, and its possible that managing/integrating it could lead to a result.

**Comment by abebeos** [comment]
> @arturhefczyc maybe you want to apply a separate bounty to #172 to solve this specific "managed k8s problem"?

@arturhefczyc , possibly it is as simple as ensuring that https://github.com/Mailu/Mailu/pull/2300 gets validated/merged

**Comment by abebeos** [comment]
@vibhuti019 , before you invest more time into this:  please note that the documentation task is nearly completed. One major barrier exists, which is the load-balancing issue or managed k8s (#172), which is most possibly fixed via **(the luckily already existent)** https://github.com/Mailu/Mailu/pull/2300. So I will at some point claim the bounty, either by providing a non-managed k8s setup, or by using the updated mailu with lb-support. Just be aware of this.

**Comment by arturhefczyc** [comment]
Thank you all for the effort and work. The most important question at this point is:
Is it doable at all?

If this is not doable, then we all did really wasted our time on this and then what would you recommend instead? Is there any other email system deploy-able on k8s?

If it is doable, then all the hours are not wasted, instead they are moving us closer to having final solution. But then what would be the best way to have it working and how much effort/cost is required? Can you give me some kind of tasks required to complete and cost estimation? Having this information I might be able to cover the development costs.

As the bountysource might be not reliable or preferred way of payment, I am OK with sending payments in a different way, like paypal perhaps.

**Comment by arturhefczyc** [comment]
The estimated costs are acceptable to me and I am happy to pay for the work. Let me know what form of payment works best for you and if you are OK with it, we can move forward with implementation.

**Comment by abebeos** [comment]
@all: To summarize: this documentation-task is currently not solvable.

The mailu k8s prerequisites can be installed using the tigase k8s-scripts (https://github.com/tigase/k8s-scripts).

One hits then on a **major** barrier: managed-k8s (like e.g. Oracle Cloud) is not supported by mailu.

* validate/update/fix tigase k8s-scripts (to simplify mailu prerequisites installation)
* provide managed k8s support (see #172)

**Comment by arturhefczyc** [comment]
This is not quite correct. 

We (our dev team) have a working Mailu deployment on managed k8s on Oracle Cloud. Still tweaking some details but it is working just fine and it is not an open relay.

All installed using the scripts set mentioned above by @abebeos.

We are polishing the setup and installation scripts and once this is done we will release thorough documentation.

That's said, we have it working on Oracle Cloud k8s only. We have not tested it anywhere else and I do not know if it will. If times allows us we will test/verify it on other providers.

**Comment by arturhefczyc** [comment]
Oh, man, are you saying that something is not possible just because you are unable to do it? That's not how logic works. On the contrary, one person doing something is a proof good enough that it can be done.

And that is my point in response to your summary that Mailu on k8s on Oracle Cloud is not possible. We did it, so it is possible. How we did it, does change the main point.

I will update the issue when we have the documentation ready for the tools we use to set everything up.

**Comment by abebeos** [comment]
> Oh, man, are you saying that something is not possible just because you are unable to do it? 

I didn't even try that much, once I got the information that you have it working. You should know this.

> That's not how logic works. On the contrary, one person doing something is a proof good enough that it can be done.

The logic is simple: A non-public result is a non-existent existent result (for OSS)

Anyone, including mailu folks, says "it is currently not possible". This is the "public status".

> We did it, so it is possible. How we did it, does change the main point.

You did it, for yourself, **non-public**, and I  (trying to fulfill a bounty to document the process) am trying for a month or so for to make you PUBLISH this work, so I can finish the docs.

(I took the effort to use your k8s-scripts for the prerequisite installation, to avoid further effort once your mailu-script gets updated)

**Comment by arturhefczyc** [comment]
Bounty is always offered to people for solving a problem. 
You want me to solve the problem, publish solution, so you can copy it and take the bounty?

**Comment by abebeos** [comment]
The bounty is not about solving the missing managed k8s support (https://github.com/Mailu/helm-charts/issues/172), but about streamlining the installation and documenting it.

You are free to ignore all the work that I've done (e.g. validating/documenting k8s prerequisites, identifying the core problem, finding possible solution-paths, invalidating them, documenting everything during processing, etc.).

There are people that do not see this as "work".

Won't fight them.

**Comment by arturhefczyc** [comment]
It's been a while but finally it is ready:

https://just-4.fun/blog/howto/oc-k8s-mailu/

It is quite long but I wanted to make it as detailed as possible. Hopefully it covers all the information necessary.

I am going to leave the ticket open for a few weeks waiting for any feedback. Once it is polished I will close the ticket.

**Comment by abebeos** [comment]
Looks good, just one thing:

Though it is nice to have such thorough documentation (like the one you provided), one part is missing:

- the setup-process itself needs to be streamlined more, to the point that one can checkout a repository, do some configs and execute a few cli commands. Whenever problems come up, one can fall-back to the full documentation.

**Comment by abebeos** [comment]
Some people will prefer this style of thorough documentation. Some others would prefer the compact style that I started.

It is usually better when one person provides the technical results, another one documents/refines/simplifies, yet another one verifies the final result (end users replaying the installation).

For the case you want to provide a 2nd set of a simplified documentation, let me know. I would be available to complete the work that I started (assuming that all technical barriers are now fixed).

**Comment by arturhefczyc** [comment]
Right now I have no idea how could I further simplify the docs. I know it is long and detailed but for me all I put there is essential information to get a working email system.

If you have an idea on how to create simplified version of it, go ahead.

**Comment by abebeos** [comment]
To clarify, this would be in context of the bounty.

I would 

* complete the work on `mailu-k8s` which got stuck with at the load-balancing issue.
* point to your full documentation for reference / problem-solving
* report any issues in your documentation that I detect during further work.

**Comment by arturhefczyc** [comment]
I am not sure if I understood it right. You want to claim the bounty for the guide?

**Comment by abebeos** [comment]
Yes, for a simplified/less verbose guide as started within https://github.com/abebeos/mailu-k8s.

(though I've not yet assessed the full effort, need to review all of your guide first)

**Comment by arturhefczyc** [comment]
The bounty is for a "Full guide on how to deploy Mailu on k8s" which I have already done.

There is no bounty for a simplified/less verbose guide. At least not such a bounty offered by myself.

**Comment by abebeos** [comment]
Ok then, all good!

**Comment by micw** [comment]
@arturhefczyc Thank you for the guide. I added a link to it to our readme.

**Comment by ayamGelo** [comment]
Should we add how to migrate data from dockerize to k8s? I think it's just data transfer to pvc

**Comment by arturhefczyc** [comment]
I have a guide for IMAP accounts migration: https://just-4.fun/blog/howto/imap-accounts-migration/
This is more reliable way to migrate data as it allows to move your mailbox between any IMAP provider.
Moving just folders from dockerized Mailu to k8s Mailu might be much faster but I was afraid that there might be some compatibility problems or so other issues with file permissions, ownership etc...
Using imapsync seems like a more reliable and easiest way.

But, feel free to add a guide for a direct data transfer if this works for you. I would be happy to add a link to the guide in my post.

**Comment by Maximilian-Maag** [comment]
currently working a mailu deployment in k8s would really like to read that guide.
